coverage/PENDING_REVIEW/intel/0450-COVERAGE-nir-lower-vars-to-ssa-571/cov-unused-access-past-matrix-elements.amber

```diff
@@ -0,0 +1,301 @@
+#!amber
+
+# Copyright 2021 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+# A test for a coverage-gap found by the GraphicsFuzz project.
+
+# Short description: A fragment shader that covers specific NIR code paths
+
+# The test passes because the shader always writes red.
+
+# Optimized using spirv-opt with the following arguments:
+# '--vector-dce'
+# '--scalar-replacement=100'
+# '--eliminate-local-multi-store'
+# '--copy-propagate-arrays'
+# '--convert-local-access-chains'
+# '--reduce-load-size'
+# '--eliminate-dead-inserts'
+# '--eliminate-dead-inserts'
+# '--ccp'
+# spirv-opt commit hash: a0370efd589be33d5d9a85cfde2f85841b3755af
+
+
+
+SHADER vertex variant_vertex_shader PASSTHROUGH
+
+# variant_fragment_shader is derived from the following GLSL:
+# #version 320 es
+# #define _int_1 _GLF_uniform_int_values[0]
+# #define _int_0 _GLF_uniform_int_values[1]
+# #define _int_2 _GLF_uniform_int_values[2]
+# #define _int_4 _GLF_uniform_int_values[3]
+# #define _float_1_0 _GLF_uniform_float_values[0]
+# #define _float_0_0 _GLF_uniform_float_values[1]
+# #define _float_3_0 _GLF_uniform_float_values[2]
+#
+# precision highp int;
+# precision highp float;
+#
+# // Contents of _GLF_uniform_int_values: [1, 0, 2, 4]
+# layout(set = 0, binding = 0) uniform buf0
+# {
+#     int _GLF_uniform_int_values[4];
+# };
+# // Contents of _GLF_uniform_float_values: [1.0, 0.0, 3.0]
+# layout(set = 0, binding = 1) uniform buf1
+# {
+#     float _GLF_uniform_float_values[3];
+# };
+# layout(location = 0) out vec4 _GLF_color;
+#
+# void main()
+# {
+#     mat4x3 m43 = mat4x3(_float_0_0);
+#     m43[_int_1][_int_1] = _float_1_0;
+#     float sums[3] = float[3](_float_1_0, _float_1_0, _float_1_0);
+#
+#     for(int i = _int_1; i < _int_4; i++)
+#     {
+#         sums[_int_1] += m43[i][_int_1];
+#     }
+#
+#     // Accessing past the matrix elements causes undefined
+#     // behavior, but the value is not used.
+#     int a = 4;
+#     sums[_int_2] += m43[a][_int_1];
```

Comment:
@paulthomson What do we think of this UB?

@Vihanakangas what if `a` were clamped to be in range? i.e. replace `a` with clamp(a, 0, 2) - would that still hit the coverage point while avoiding the UB?

Reply:
The coverage point was hit only if it was very clear for the copiler that an OOB access was going to happen.

Reply:
If this line does actually execute (and I think it does) then I don't think we should add this test. Can you try making it so this line does not actually get executed (but the compiler wouldn't know)? That way, the cov point might still trigger, as the OOB access is still obvious. We had a case like this before and we agreed that it should be allowed, as long as the access (in the SPIR-V) never actually executes. You could try making it so the `int a = 4;` line also doesn't execute.